### PR TITLE
refactor: Clarify Room.join() behavior and make sure DM link is purged if room not found

### PR DESCRIPTION
**Pull request by krille-chan**

### CODE REVIEW

File: lib/src/room.dart [hunk 1]
```diff
@@ -1150,21 +1150,29 @@ class Room {
...
-      final dmId = directChatMatrixID;
-      if (dmId != null) {
-        await addToDirectChat(dmId);
-      }
+      if (dmId != null) await addToDirectChat(dmId);
 
       // now join
       await client.joinRoomById(id);
     } on MatrixException catch (exception) {
+      if (dmId != null) await removeFromDirectChat();
```
Comment (td-famedly) [inline lib/src/room.dart:1167]: Can we put this in the leave room call?

Comment (td-famedly) [reply lib/src/room.dart:1167]: The spec doesn't say so, but it is intuitive to update the m.direct list if you leave the room right? Or does this have any edge cases?

Comment (krille-chan) [reply lib/src/room.dart:1167]: We had this before and removed by a reason that in the archive we still want to see that this was once a DM room with this user.

Comment (td-famedly) [reply lib/src/room.dart:1167]: ah right! thanks

**Review by td-famedly — APPROVED** [review]
lg, I don't see `M_NOT_FOUND` used for join events in the spec so this should be fine

**Pull request merged**